core/src/main/scala/org/apache/spark/deploy/master/ui/MasterWebUI.scala

```diff
@@ -49,6 +56,23 @@ class MasterWebUI(
       "/app/kill", "/", masterPage.handleAppKillRequest, httpMethods = Set("POST")))
     attachHandler(createRedirectHandler(
       "/driver/kill", "/", masterPage.handleDriverKillRequest, httpMethods = Set("POST")))
+    attachHandler(createServletHandler("/workers/kill", new HttpServlet {
+      override def doPost(req: HttpServletRequest, resp: HttpServletResponse): Unit = {
+        val hostPorts: Seq[String] = Option(req.getParameterValues("host"))
+          .getOrElse(Array[String]()).map(_.toLowerCase(Locale.ROOT)).toSeq
+        logInfo(s"Received request to decommission workers $hostPorts from ${req.getRemoteAddr}")
+        if (!isDecommissioningRequestAllowed(req)) {
+          resp.sendError(HttpServletResponse.SC_METHOD_NOT_ALLOWED)
+        } else {
+          val removedWorkers = masterEndpointRef.askSync[Integer](DecommissionHostPorts(hostPorts))
+          if (removedWorkers <= 0) {
```

Comment:
when would the `removedWorkers < 0` happen?

Reply:
Never :-P ... I will handle that as an error separately.